README.md
NetCDF-Decoders
===================
Greetings!  This file briefly describes the Unidata netCDF Decoders package.

###Table of Contents
* [Notice of Deprecation](#Notice_of_Deprecation)
* [Introduction](#Introduction)
* [Installation](#Installation)
* [Support & Mailing List Archives](#Support)


### <a name="Notice_of_Deprecation"></a>Notice of Deprecation
The netCDF Decoders package is **no longer being actively maintained or supported** by the Unidata Program Center. 
### <a name="Introduction"></a>Introduction

The netCDF Decoders convert Textual Reports to netCDF format. To convert both GRIB 1 and GRIB 2 data, use the [netCDF Java package](http://www.unidata.ucar.edu/software/thredds/current/netcdf-java/)


Reference the [RELEASE-NOTES](src/RELEASE-NOTES) for version-specific information.


### <a name="Installation"></a>Installation

See the file [INSTALL](src/INSTALL) in the top-level src directory of the netCDF Decoders distribution for installation instructions.

If needed, reference the [CUSTOMIZE](src/CUSTOMIZE) file for instructions on how to customizing the configuation process.


### <a name="Support"></a>Support & Mailing List Archives

**Support Archives**

The Unidata Program Center is **no longer providing support** for the netCDF Decoders  package.

However, you can also view past [netCDF Decoders support questions]( http://www.unidata.ucar.edu/support/help/MailArchives/decoders/) answered by Unidata developers in an archive maintained by Unidata as a courtesy to our users.

**Defunct Mailing List**

The netCDF Decoders had a community mailing list associated with it at one time. The mailing list is **no longer active**, but the [list archive](http://www.unidata.ucar.edu/mailing_lists/archives/decoders/) is still available for reference.



src/RELEASE-NOTES
$Id: RELEASE-NOTES,v 1.37 2004/12/21 20:55:26 rkambic Exp $

Release	Date		Notes
_________________	______________________________________________________
5.0.2 2013-05-07	Corrected gfs_37-44.cdl (grids are 1.25x1.25 degree);
                        added ukmet_21-24.cdl (like GFS AVN, these are 1.25x1.25 degree
                        grids); added '-n' flag to gribtonc (says to not expand
                        the record dimension past what is defined in the CDL/existing
                        netCDF file); updated 'port' subdirectory with latest code
                        (so that the package will build on CentOS 6.4 x86_64)
5.0.1 2012-06-13	Remade ruc2_236.cdl because of transfer of the RUC2 production
			from FSL to RAP 
5.0.0 2007-10-10	The perl Decoders have been moved to their own package
			so the process of deprecating this package can take place.
			This package is being deprecated because of it's 
			limitation of not being able to decode Grib2, compile
			problems, and static parameter table.
4.1.7 2007-09-27	metar2nc decoders removes reports with illegal cntl chars
			changed dist from Z to gz format
4.1.6 2007-08-20	fixed typo in eta_211.cdl
4.1.5 2007-02-09	added a -c option to metar2nc* so it can process archive
                        data from the command line.
4.1.4 2006-10-04	fixed bug in ua2nc, putting 23hr data into wrong file.
4.1.3 2006-09-14	fixed bug in ua2nc, putting 23hr data into wrong file.
4.1.2 2006-09-06	added more levels
4.1.1 2006-08-23	minor bug fixes, recs.c added import, metar2nc.new1
			SPD and GUST are now floats when converted from knots
			to m/s. etc
4.0.0 2005-12-19	update the METAR table sfmetar_sa.tbl, created new
			metar.cdl.new1, new table lsfstns.tbl for synoptic 
			decoder, and new table for upperair decoder snstns.tbl
			updated other cdls. the new table names cause mods to
			syn2nc and ua2nc.
4.0.0 2005-10-05	changed perl from /usr/local/bin/perl to /usr/bin/perl
			pqact entries don't need yyyymm argument anymore.
			new naming convention, ie surface_METAR_yyyymmdd_hhmm.nc
			added -h option to write hourly files for the all perl
			decoders, default is daily file. updated sfmetar.tbl, 
			snworld.tbl and systns_upc.tbl. fixed bug in syn2nc 
			that returned wind DIR greater than 900, removed java 
			Grib decoder jars, it's has it own release package now
			called grib-decoder-java_2.2.10.tar.gz. gfs_211.cdl,
			eta_211.cdl, and ruc2_236.cdl were updated to add
			new parmeters. managed log files to only keeps 3 logs
			added global attributes to metar.cdl.new1
			changed testdata.wmo to testdata.wmo.bak, etc

3.1.1 2005-05-24	changed units of Dx and Dy to be meters instead of km 
			in all cdls. changed cdl version to be 1.0
3.1.0 2005-05-24	fixed the perl decoders so unicode characters don't put
			decoder into infinite loop. This fix is needed for 
			perl version 5.8.0 and higher. fixed buoy2nc and syn2nc
			where min was set to 60 and timegm() was called. 
			Also added the java grib.jar to the release.

3.0.9 2005-03-01	fixed the perl decoder cdls to be udunits compliant.  
3.0.8 2005-02-04	fixed metar2nc and metar2nc.new1 dew point temperature
			incase the coding changes from /M00 to /M0
3.0.7 2004-12-21	added level 245, changed files: levels.h gribtocdl.c 
			levels.c nc.c
3.0.6 2004-10-21	some parameters started with a number, thats illegal
			in netcdf. also correct units for oztot (ozone total)
3.0.5 2004-10-18	update cdls and new version of metar.new1, metar.cdl.new

3.0.4 2004-07-22	rewrote perl decoders to be more robust especially
			when checking the report time for accuracy. Changed
			test to read in more stations from the sfmetar_sa.tbl
			station file. Added metar2nc.new1 and metar.cdl.new1 as
			a new perl decoder, new netCDF format. New script
			makeCDL1.pl makes a cdl file metar.cdl.new1. In recs.c
			upped RECS_INIT_SIZE to 64. Updated cdls
			In nc.c added stronger test to parmname, levels
			mods to test.pl for real time data, mods to
			src/*/testdata.wmo for real time testing. Fixed 
			return value for time range indicator in gribtocdl

3.0.3 2004-03-12	added Writer scripts into all the perl decoders. These
			scripts bust-up bulletins into individual reports
			added missing level info to gribtocdl.c program
			added Makefile.in to zone MANIFEST
			Fixed century bug, files recs.c units.c gribdump

3.0.2 2004-02-01	change #define MAX_GRIB_SIZE 200000
			added forecasttime variable to gribtocdl
			rename all cdls to reflect new model names
			updated INSTALL and CUSTOMIZE files
			in cdl dir, updated Makefile.in and README
			mods to metar2nc, made matches more robust

3.0.1 2003-04-25	added CF standard_name, forecasttime, and enhanced
			Conventions attribute to acknowledge CF conventions
			to all model cdls. Modified recs.c to write out
			forecasttime.

3.0.0 2003-03-25	added dimension valtime_offset for the valtime_offset
			variable. Enhanced cdl to have the data for the 
			valtime_offset var.  This permits the data to be stored
			in monotonic increasing order.  Added parameter_number
			and level_number attributes to variables.

2.4.4 2003-02-25        added vars gust_sfc to ruc2.cdl and vars Tmin_fhg and
			Tmax_fhg to avn-x.cdl
			mods to gdes.c gribtocdl.[1c] inetutil.c ulog.h so 
			code compiles on AIX boxes.

2.4.3 2003-02-21        modifications so user parms functionality works
			correctly in the code, new functionality to 
			gribdump, also updated the man pages. files:
			gribtocdl.c, params.c, gribdump.c, gribtonc.1 and
			gribdump.1
			
2.4.2 2003-02-19        Added the PNO indicator for Metar decoder
			In gdes.c, changed i_dim from lat to lon
			fixed avn-1.0x1.0.cdl also

2.4.1 2003-01-31        Fixed some error for the avn-1.0x1.0 Grid.  Also fixed
			errors for the Gausian lat/lon grid, plus other fixes
			mods to gribtocld.c gdes.c nc.c 

2.4.0 2002-11-04        Merge in code made by Arild Burud and Terje Reite for

src/gribtonc/nuwg.c
/*
 *   Copyright 1995, University Corporation for Atmospheric Research.
 *   See ../COPYRIGHT file for copying and redistribution conditions.
 */
/* $Id: nuwg.c,v 1.13 2003/03/25 19:47:49 rkambic Exp $ */


/*
 * Permits some customization of which netCDF names are used in netCDF files
 * for gridded data that follow the NUWG convention.  Although a specific
 * default name is used, this interface allows a user to check other names
 * or aliases for NUWG dimension or variable names.  Ultimately this should
 * be table-driven, with the table parsed to initialize the tests done here,
 * so that recompilation is not necessary to change names.
 */


#include <netcdf.h>
#include "ulog.h"
#include "nuwg.h"


/*
 * Returns name for specified NUWG component, or -1 if there is no
 * such component.
 */
char *
nuwg_name(which)
    enum ncpart which;
{
    switch(which) {
    case DIM_FHG:
	return "fhg";
    case DIM_LEVEL:
	return "level";
    case DIM_NAV:
	return "nav";

    case DIM_SIGMA:
	return "sigma";
    case DIM_HY:
	return "hyb";
    case DIM_FH:
	return "fh";
    case DIM_BLS:
	return "bls";
    case DIM_ISEN:
	return "isen";
    case DIM_PDG:
	return "pdg";
    case DIM_FHGH:
	return "fhgh";
    case DIM_DBS:
	return "dbs";
    case DIM_FL:
 	return "fl";

    case DIM_LBLS:
	return "lbls";
    case DIM_LFHG:
	return "lfhg";
    case DIM_LFHM:
	return "lfhm";
    case DIM_LHY:
	return "lhy";
    case DIM_LISEN:
	return "lisen";
    case DIM_LISH:
	return "lish";
    case DIM_LISM:
	return "lism";
    case DIM_LISO:
	return "liso";
    case DIM_LPDG:
	return "lpdg";
    case DIM_LSH:
	return "lsh";
    case DIM_LS:
	return "ls";

    case DIM_NGRIDS:
	return "ngrids";
    case DIM_VALOFFSETS:
	return "valtime_offset";
    case VAR_REFTIME:
	return "reftime";
    case VAR_VALTIME:
	return "valtime";
    case VAR_DATETIME:
	return "datetime";
    case VAR_FORECASTTIME:
	return "forecasttime";
    case VAR_VALOFFSET:
	return "valtime_offset";
    case VAR_GRID_TYPE_CODE:
	return "grid_type_code";
    case VAR_GRID_CENTER:
	return "grid_center";
    case VAR_GRID_NUMBER:
	return "grid_number";
    case VAR_RESCOMP:
	return "ResCompFlag";
    case VAR_MODELID:
	return "model_id";
    case VAR_DI:
	return "Di";
    case VAR_DJ:
	return "Dj";
    case VAR_DX:
	return "Dx";
    case VAR_DY:
	return "Dy";
    case VAR_J:
	return "J";
    case VAR_K:
	return "K";
    case VAR_LA1:
	return "La1";
    case VAR_LA2:
	return "La2";
    case VAR_LAP:
	return "Lap";
    case VAR_LATIN:
	return "Latin";
    case VAR_LATIN1:
	return "Latin1";
    case VAR_LATIN2:
	return "Latin2";
    case VAR_LO1:
	return "Lo1";
    case VAR_LO2:
	return "Lo2";
    case VAR_LOP:
	return "Lop";
    case VAR_LOV:
	return "Lov";
    case VAR_M:
	return "M";
    case VAR_MODE:
	return "Mode";
    case VAR_N:
	return "N";
    case VAR_NI:
	return "Ni";
    case VAR_NJ:
	return "Nj";
    case VAR_NR:
	return "Nr";
    case VAR_NX:
	return "Nx";
    case VAR_NY:
	return "Ny";
    case VAR_ORIENTATION:
	return "Orientation";
    case VAR_PROJFLAG:
	return "ProjFlag";
    case VAR_ROTANGLE:
	return "RotAngle";
    case VAR_ROTLAT:
	return "RotLat";
    case VAR_ROTLON:
	return "RotLon";
    case VAR_SPLAT:
	return "SpLat";
    case VAR_SPLON:
	return "SpLon";
    case VAR_STRETCHFACTOR:
	return "StretchFactor";
    case VAR_STRETCHLAT:
	return "StretchLat";
    case VAR_STRETCHLON:
	return "StretchLon";
    case VAR_TYPE:
	return "Type";
    case VAR_XO:
	return "Xo";
    case VAR_XP:
	return "Xp";
    case VAR_YO:
	return "Yo";
    case VAR_YP:
	return "Yp";
    default:
	uerror("nuwg_name() called for bad component: %d", which);
	break;
    }
    /* default */
    return 0;
}


/*
 * Returns dimension ID for specified NUWG dimension, or -1 if there is no
 * such dimension.
 */
int
nuwg_getdim(ncid, which)
    int ncid;
    enum ncpart which;
{
    return ncdimid(ncid, nuwg_name(which));
}


/*
 * Returns variable ID for specified NUWG variable, or -1 if there is no
 * such variable.
 */
int
nuwg_getvar(ncid, which)
    int ncid;
    enum ncpart which;
{
    return ncvarid(ncid, nuwg_name(which));
}
